### PR TITLE
update(transformers):  general clean up & improvment

### PR DESCRIPTION
**Pull request by ajweeks**
This change fixes typos, unifies variable naming, and improves wording of several sections.

Note that I wasn't able to get the scripts (nor precommit hooks) running on Windows so I have edited each file manually.

### CODE REVIEW

File: workshops/transformer/transformer.ipynb [hunk 1]
```diff
@@ -59,7 +59,7 @@
                 "> ##### Learning objectives\n",
                 ">\n",
                 "> - Understand what a transformer is used for\n",
-                "> - Understand causal attention, and what a transformer's output representsalgebra operations on tensors\n",
+                "> - Understand causal attention, and what a transformer's output represents (algebraic operations on tensors)\n",
```
Comment (ddorn) [inline workshops/transformer/transformer.ipynb:62]: I'm can't make sense of this sentence (nor of the previous one). I think it might be two different goals (1. understand attn and outputs; 2. understand operations on tensors)

File: workshops/transformer/transformer.ipynb [hunk 2]
```diff
@@ -349,7 +349,7 @@
                 "\n",
                 "#### **Step 1:** Convert text to tokens\n",
                 "\n",
-                "The sequence gets tokenized, so it has shape `[batch, seq_len]`. Here, the batch dimension is just one (because we only have one sequence).\n"
+                "The sequence gets tokenized, so it has shape `[batch, seq_len]`. Here, the batch dimension is just one because we only have one sequence. Think of `batch` as how many sentences we process, and `seq_len` as sentence length in tokens.\n"
```
Comment (ddorn) [inline workshops/transformer/transformer.ipynb:352]: Maybe use "documents" instead of sentences, to be closer to say something true?

File: workshops/transformer/transformer.ipynb [hunk 3]
```diff
@@ -211,7 +211,7 @@
             "source": [
                 "### Converting sub-units to vectors\n",
                 "\n",
-                "We basically make a massive lookup table, which is called an **embedding**. It has one vector for each possible sub-unit of language we might get (we call this set of all sub-units our **vocabulary**). We label every element in our vocabulary with an integer (this labelling never changes), and we use this integer to index into the embedding.\n",
+                "Text is split into small pieces called tokens (e.g., words or word parts) using a tokenizer. Each token gets a number, which we turn into a vector (a list of numbers) using a table called an **embedding**.\n",
```
Comment (ddorn) [inline workshops/transformer/transformer.ipynb:214]: I prefer your order of descriptions, which follow the order of the steps, but I prefer the details of the pas version, which details more what an embedding is.

File: workshops/transformer/transformer.ipynb [hunk 4]
```diff
@@ -629,7 +629,7 @@
             "source": [
                 "### MLP\n",
                 "\n",
-                "The MLP layers are just a standard neural network, with a singular hidden layer and a nonlinear activation function. The exact activation isn't conceptually important ([GELU](https://paperswithcode.com/method/gelu) seems to perform best).\n",
+                "The MLP layers are just a standard machine learning, with a singular hidden layer and a nonlinear activation function. The exact activation function used isn't conceptually important, though [GeLU](https://paperswithcode.com/method/gelu) is often chosen for its performance.\n",
```
Comment (ddorn) [inline workshops/transformer/transformer.ipynb:632]: "neural network" was the right term here. I also preferred the wording in the original sentence: "seems to perform best" indicates more clearly that we have no idea why, and it makes it less ambiguous about which performance we are discussing (that is, the performance of the function that requires less computation, or the performance of the model on the test set).

File: workshops/transformer/transformer.ipynb [hunk 5]
```diff
@@ -667,9 +667,10 @@
                 "\n",
                 "#### LayerNorm\n",
                 "\n",
-                "* Simple normalization function applied at the start of each layer (i.e. before each MLP, attention layer, and before the unembedding)\n",
-                "* Converts each input vector (independently in parallel for each batch x position residual stream vector) to have mean zero and variance 1.\n",
-                "* Then applies an elementwise scaling and translation\n",
+                "* Forces vectors into the range `[0, 1]` which helps the model learn more efficiently.\n",
```
Comment (ddorn) [inline workshops/transformer/transformer.ipynb:670]: This seems wrong, or very interpretation dependent. The elements of the vector are not sent to `[0, 1]`, nor is the norm. The norm is scaled to precisely one then scaled (as said later). 
Maybe worth of note here is that vector norm 1 and variance 1 is the same thing when the mean is 0.

File: workshops/transformer/transformer.ipynb [hunk 6]
```diff
@@ -679,13 +680,12 @@
                 "#### Positional embeddings\n",
                 "\n",
                 "* **Problem:** Attention operates over all pairs of positions. This means it's symmetric with regards to position - the attention calculation from token 5 to token 1 and token 5 to token 2 are the same by default\n",
-                "    * This is dumb because nearby tokens are more relevant.\n",
-                "* There's a lot of dumb hacks for this.\n",
-                "* We'll focus on **learned, absolute positional embeddings**. This means we learn a lookup table mapping the index of the position of each token to a residual stream vector, and add this to the embed.\n",
+                "    * This is inefficient because nearby tokens are often more relevant.\n",
+                "* One solution is **learned, absolute positional embeddings**. This involves a learned lookup table mapping each token's position to a residual stream vector, and adding this to the embed.\n",
```
Comment (ddorn) [inline workshops/transformer/transformer.ipynb:684]: Thank you for removing the "this is dumb" !!

One thing maybe: it's not only that it's inefficient, it's also that the model has no easy way of knowing where tokens are in the text.

File: workshops/transformer/transformer.ipynb [hunk 7]
```diff
@@ -1149,15 +1149,15 @@
...
+                "    \"???,??? -> batch token_Q head d_head\",\n",
                 ")\n",
                 "\n",
                 "# Calculate output (by applying matrix W_O and summing over heads, then adding bias b_O)\n",
                 "attn_out = (\n",
                 "    einops.einsum(\n",
                 "        z,\n",
                 "        self.W_O,\n",
-                "        \"batch posn_Q nheads d_head, nheads d_head d_model -> batch posn_Q d_model\",\n",
+                "        \"batch token_Q head d_head, head d_head d_model -> batch token_Q d_model\",\n",
```
Comment (ddorn) [inline workshops/transformer/transformer.ipynb:1160]: Thank you so much for all those fixes!

**Review by ddorn — COMMENTED** [review]
Thanks for going through the notebook and fixing so many small things. I've added a few comments, because I think a few things can be improved.
You can resolve the comments once you're done, and leave me a message for those you don't agree with and then I'll merge ;)

Also I might have fixed the bug on master that prevented you from running the pre-commit, so you should pull and try again! (so that you don't have to copy paste painfully all your changes...)

**Comment by ajweeks** [comment]
Updated with your feedback!

File: workshops/transformer/transformer.ipynb [hunk 8]
```diff
@@ -369,10 +337,10 @@
             "cell_type": "markdown",
             "metadata": {},
             "source": [
-                "#### **Step 2:** Map tokens to logits\n",
+                "#### **Step 2:** Feed tokens to model\n",
```
Comment (ajweeks) [inline workshops/transformer/transformer.ipynb:340]: Still not 100% happy with this phrasing but "map" seems too general for describing the entire feed forward pass?

File: workshops/transformer/transformer.ipynb [hunk 9]
```diff
@@ -306,37 +257,54 @@
...
                 "\n",
-                "\n",
-                "\n",
                 "### Some tokenization annoyances\n",
                 "\n",
-                "There are a few funky and frustrating things about tokenization, which causes it to behave differently than you might expect. For instance:\n",
+                "There are a few funky and frustrating things about tokenization which cause it to behave differently than you might expect, such as how it handles capitalization and leading spaces. \n",
                 "\n",
-                "#### Whether a word begins with a capital or space matters!\n"
+                "You can spend a minute playing around with tokenization in [tiktokenizer](https://tiktokenizer.vercel.app/?model=gpt2). Enter the word \"Ralph\", then see how tokenization changes with an initial capital letter vs. not, and with a leading space vs. not. Then compare that with how a common word like \"water\" gets tokenized in different forms."
```
Comment (ajweeks) [inline workshops/transformer/transformer.ipynb:264]: I personally find this website more instructive than running cells here because of the instant feedback. Thoughts?

**Comment by ajweeks** [comment]
How are fixes normally propagated to the arena notebook?

File: workshops/transformer/transformer.ipynb (hunk 9 above)
Comment (ddorn) [reply workshops/transformer/transformer.ipynb:264]: Seems reasonable, I'm happy to have this.

File: workshops/transformer/transformer.ipynb (hunk 8 above)
Comment (ddorn) [reply workshops/transformer/transformer.ipynb:340]: I'm happy with `map`, or `transform`. Here we view the transformer as a sequence of steps, each step being a function that maps inputs to outputs. There's more information in the original "map tokens to logits" than your proposition.

**Comment by ajweeks** [comment]
Great, then I think this is ready for merging now :)
